src/python/pants/backend/go/target_type_rules.py

```diff
@@ -119,27 +151,63 @@ async def infer_go_dependencies(
-            # import path. This may be the case, but bears confirmation.
-            #
-            # TODO: Also cleanup mandatory go_module ownership, so module_import_path is not Optional[str].
-            if module_import_path and import_path.startswith(module_import_path):
-                inferred_dependencies.append(address)
-                found_module_import = True
-
-        if found_module_import:
+        # Infer third-party dependencies on _go_ext_mod_package targets.
+        candidate_third_party_packages = package_mapping.mapping.get(import_path, ())
+        if len(candidate_third_party_packages) > 1:
+            logger.warning(
+                f"Ambiguous mapping for import path {import_path} on packages at addresses: {candidate_third_party_packages}"
+            )
+        elif len(candidate_third_party_packages) == 1:
+            inferred_dependencies.append(candidate_third_party_packages[0])
+        else:
+            logger.debug(
+                f"Unable to infer dependency for import path '{import_path}' "
+                f"in go_package at address '{this_go_package.address}'."
+            )
+
+    return InferredDependencies(inferred_dependencies, sibling_dependencies_inferrable=False)
+
+
+class InjectGoExternalPackageDependenciesRequest(InjectDependenciesRequest):
+    inject_for = GoExtModPackageDependencies
+
+
+# TODO: This duplicates first-party dependency inference but that other rule cannot operate on _go_ext_mod_package
+# targets since there is no sources field in a _go_ext_mod_package. Consider how to merge the inference/injection
+# rules into one.
+@rule
+async def inject_go_external_package_dependencies(
+    request: InjectGoExternalPackageDependenciesRequest,
+    goroot_imports: ResolvedImportPathsForGoLangDistribution,
+    package_mapping: GoImportPathToPackageMapping,
+) -> InjectedDependencies:
+    this_go_package = await Get(
+        ResolvedGoPackage, ResolveExternalGoPackageRequest(request.dependencies_field.address)
+    )
+
+    # Loop through all of the imports of this package and add dependencies on other packages and
+    # external modules.
+    inferred_dependencies = []
+    for import_path in this_go_package.imports + this_go_package.test_imports:
+        # Check whether the import path comes from the standard library.
+        if import_path in goroot_imports.import_path_mapping:
             continue
 
-        logger.debug(
-            f"Unable to infer dependency for import path '{import_path}' "
-            f"in go_package at address '{this_go_package.address}'."
-        )
+        # Infer third-party dependencies on _go_ext_mod_package targets.
+        candidate_third_party_packages = package_mapping.mapping.get(import_path, ())
+        if len(candidate_third_party_packages) > 1:
+            logger.warning(
+                f"Ambiguous mapping for import path {import_path} on packages at addresses: {candidate_third_party_packages}"
+            )
```

Comment:
Ditto on adding a TODO to use the standardized ambiguity stuff.

Reply:
TODO